### PR TITLE
Patch API integration changes

### PR DESCRIPTION
**Pull request by maier49**
**Type:** bug / feature

**Description:** 

Updates the `patch()` API with improvements based on work integrating stores into TodoMVC
Dependent on changes in https://github.com/dojo/stores/pull/32.

This makes the patch API easier to use by allowing objects to be passed and treated as 'partial' state. There are two potential gotchas with these changes:
If id is a property on the object that should be updated, and should not be treated as the store id, then the id will need to be passed in the options(requiring a single argument patch), or the `Map` signature will need to be used. Otherwise the id property will be treated as the unique id of the item and will not be modified by the operation.

The other case is if the store items contain a `patch` property with this structure:

```
{
  apply: Function,
  operations: Array,
  toString: Functoin
}
```

If this is the case then the `Map` signature should be used to update the items. The alternative is to strictly enforce the usage of the `{ id: string; patch: Patch }` format for updates, which in most cases is less convenient for the user but prevents these edge cases from being issues. 

Please review this checklist before submitting your PR:
- [ ] There is a related issue
- [x] All contributors have signed a CLA
- [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
- [x] The code passes the CI tests
- [x] Unit or Functional tests are included in the PR
- [x] The PR increases or maintains the overall unit test coverage percentage
- [ ] The code is ready to be merged

### CODE REVIEW

**Comment by kitsonk** [comment]
This seems to include PR #32.  If so, let' get that merged first and this rebased.

File: src/store/createStore.ts [hunk 1]
```diff
@@ -35,7 +37,11 @@ export interface UpdateResults<T> {
 	type: StoreOperation;
 }
 
-export type PatchArgument<T> = Map<string, Patch<T, T>> | { id: string; patch: Patch<T, T> } | { id: string; patch: Patch<T, T> }[];
+export type PatchArgument<T> = Map<string, Patch<T, T>> |
+	{ id: string; patch: Patch<T, T> } |
+	{ id: string; } |
+	{ id: string; patch: Patch<T, T> }[] |
+	{id: string; }[];
```
Comment (agubler) [inline src/store/createStore.ts]: formatting

File: src/patch/createPatch.ts [hunk 2]
```diff
@@ -41,8 +41,14 @@ function _diff(from: any, to: any, startingPath?: JsonPointer): Operation[] {
...
-export function diff<T, U>(from: T, to: U): Patch<T, U> {
-	return <Patch<T, U>> createPatch(_diff(from, to));
+export function diff<T>(to: T): Patch<any, T>;
+export function diff<T, U>(from: T, to: U): Patch<T, U>;
+export function diff(from: any, to?: any) {
+	if (!to) {
+		to = from;
+		from = {};
+	}
+	return createPatch(_diff(from, to));
```
Comment (agubler) [inline src/patch/createPatch.ts]: This confused my slightly, with the switching of the arguments like that...

Wouldn't this be clearer and prevent the need to swap values? I appreciate you would need to update all current consumers of this `diff` function as I am swapping the order of the arguments.

```ts
export function diff<T>(to: T): Patch<any, T>;
export function diff<T, U>(to: T, from: U): Patch<T, U>;
export function diff(to: any, from: any = {}) {
	return createPatch(_diff(from, to));
}
```

Comment (maier49) [reply src/patch/createPatch.ts]: In my experience the `from` usually comes first. I'm drawing from a limited sample, git and a couple diff libraries, but I would find `diff(to, from)` confusing. Would having a separate method for the single argument case help resolve the confusion?

Comment (agubler) [reply src/patch/createPatch.ts]: I don't actually disagree that logically `from` Is normally first but both the API and the implementation confused me with swapping the order of the args and then depending on the order swapping the allocation in the impl.

Comment (maier49) [reply src/patch/createPatch.ts]: Fair enough. I swapped the order.

**Review by agubler — APPROVED** [review]

**Review by kitsonk — APPROVED** [review]
👍

**Pull request merged**